src/main/scala/mesosphere/marathon/core/election/CuratorElectionStream.scala

```diff
@@ -0,0 +1,243 @@
+      zooKeeperLeaderPath: String,
+      hostPort: String,
+      sq: SourceQueueWithComplete[LeadershipState]) extends Cancellable {
+
+    val client = clientCloseable.closeable
+    private lazy val leaderHostPortMetric: Timer = Metrics.timer(ServiceMetric, getClass, "current-leader-host-port")
+    private val curatorLeaderLatchPath = zooKeeperLeaderPath + "-curator"
+    private lazy val latch = new LeaderLatch(client, curatorLeaderLatchPath, hostPort)
+    private var isStarted = false
+    @volatile private var _isCancelled = false
+
+    /* Long-poll trampoline-style recursive method which calls emitLeader() each time it detects that the leadership
+     * state has changed.
+     *
+     * Given instance A, B, C, Curator's Leader latch recipe only provides A the ability to be notified if it gains or
+     * loses leadership, but not if leadership transitions between B and C. This method allows us to monitor any change
+     * in leadership state.
+     *
+     * It is important that we re-register our watch _BEFORE_ we get the current leader. In Zookeeper,
+     * watches are one-time use only.
+     *
+     * The timeline of events looks like this
+     *
+     * 1) We register a watch
+     * 2) We query the current leader
+     * 3) We receive an event (indicating child removal / addition)
+     * 4) Repeat step 1
+     *
+     * By re-registering the watch before querying the state, we will not miss out on the latest leader change.
+     *
+     * We also have a retry and (very simple) back-off mechanism. This is because Curator's leader latch creates the
+     * initial leader node asynchronously. If we poll for leader information before this background hook completes, then
+     * a KeeperException.NoNodeException is thrown (which we handle, and retry)
+     */
+    def longPollLeaderChange(retries: Int = 0): Unit = singleThreadEC.execute { () =>
+      try {
+        if (latch.getState == LeaderLatch.State.STARTED)
+          client.getChildren
+            .usingWatcher(new CuratorWatcher {
+              override def process(event: WatchedEvent): Unit =
+                if (!_isCancelled) longPollLeaderChange()
+            })
+            .forPath(curatorLeaderLatchPath)
+        emitLeader()
+      } catch {
+        case ex: KeeperException.NoNodeException if retries < 100 =>
+          // Wait for node to be created
+          logger.info("retrying")
+          Thread.sleep(retries * 10L)
+          longPollLeaderChange(retries + 1)
+        case ex: Throwable =>
+          sq.fail(ex)
+      }
+    }
+
+    /**
+      * Emit current leader. Does not fail on connection error, but throws if multiple election candidates have the same
+      * ID.
+      */
+    private def emitLeader(): Unit = {
+      val currentLeader = leaderHostPortMetric.blocking {
+        val participants = try {
+          if (client.getState == CuratorFrameworkState.STOPPED)
+            Nil
```

Comment:
Because `Nil` list will give an empty iterator? 🤔  Interesting

Reply:
Yeah, we presume no known participants :) This is translated to `Standby(None)` (we're not leader, and we don't know who is)